compiler/compiler.go

```diff
@@ -318,7 +320,16 @@ func WritePkgCode(pkg *Archive, dceSelection map[*Decl]struct{}, gls goLinknameS
 			// This decl is referenced by a go:linkname directive, expose it to external
 			// callers via $linkname object (declared in prelude). We are not using
 			// $pkg to avoid clashes with exported symbols.
-			code := fmt.Sprintf("\t$linknames[%q] = %s;\n", d.LinkingName.String(), d.Vars[0])
+			var code string
+			if recv, method, ok := d.LinkingName.IsMethod(); ok {
+				if strings.HasPrefix(recv, "*") {
+					code = fmt.Sprintf("\t$linknames[%q] = $unsafeMethodToFunction(%v,%q,true);\n", d.LinkingName.String(), d.NamedRecvType, method)
+				} else {
+					code = fmt.Sprintf("\t$linknames[%q] = $unsafeMethodToFunction(%v,%q,false);\n", d.LinkingName.String(), d.NamedRecvType, method)
+				}
```

Comment:
I think this can be further simplified to this:
```suggestion
				code = fmt.Sprintf("\t$linknames[%q] = $unsafeMethodToFunction(%v,%q,%t);\n", d.LinkingName.String(), d.NamedRecvType, method, strings.HasPrefix(recv, "*"))
```